lib/elixir/lib/kernel.ex

```diff
@@ -2442,18 +2442,36 @@ defmodule Kernel do
 
   """
   defmacro first .. last do
-    case is_float(first) or is_float(last) or
+    # require Logger
+    # Logger.debug inspect [first, last]
+    case is_tuple(first) and tuple_size(first) == 2 or
+         is_tuple(last) and tuple_size(last) == 2 or
+         is_float(first) or is_float(last) or
          is_atom(first) or is_atom(last) or
-         is_binary(first) or is_binary(last) do
+         is_binary(first) or is_binary(last) or
+         is_list(first) or is_list(last)
+         do
       true ->
-        raise ArgumentError,
-          "ranges (left .. right) expect both sides to be integers, " <>
-          "got: #{Macro.to_string({:.., [], [first, last]})}"
+        range_error(first, last)
       false ->
+        for val <- [first, last] do
+          case val do
```

Comment:
It feels too much restrictive to have this check, might be hard to consider every possible case.
Maybe just adding `is_list/1` check is a way to go?